BP-121 receive push notifications

julianweise opened the pull request:
Implemented NotificationService in order to be notified by the backend in case of any data change (Updated/New events, products, eventTypes).
Used WebSocket-Technology to connect to the backend-interface.
Due tot the fact that Jest does not support WebSocket, we are not able to achieve 100% test-coverage for the NotificationService.js.

julianweise committed:
Implemented Notification Service

In order to update the Dashboard after changes in the backend, we introduced an Notification Service as Observer-Pattern.

julianweise committed:
Notifications view

Introduced an Notification view to display occuring notifications.

julianweise committed:
Integrated Notification view

IMported Notification view into App.js

julianweise committed:
Improved data mapping of backend data

Improved mapping of data from backend in order to present information of product families in its products.

julianweise committed:
Improved style of notifications

julianweise committed:
Created NotificationView

julianweise committed:
Integrated Notificationhandling

Integrated Notification handling and forwarding in index.js in order to display notifications thrown from NotificationService in App.js

julianweise committed:
Code refactoring NotificationService

julianweise committed:
Introduced NotificationService to DashboardView

julianweise committed:
Initilization of NotificationService

+ connection data redirection

julianweise committed:
Introduced new state

Integrated new possible state of products into the Dashboard and Product-View.

julianweise committed:
Resolved react key error

julianweise committed:
(Un)registration of DashboardView for NotificationService

julianweise committed:
Stylesheet refactoring

Removed units in case of value zero.

julianweise committed:
Integrated NotificationService into ProductView

ProductView is now able to be notified about changed data in backend.

julianweise committed:
Bug wrong this context solved

julianweise committed:
Adapted tests 

Adapted tests to check new desired behavior.

julianweise committed:
Adapted mock & view to new data mapper results

julianweise committed:
Included fallback to npm WebSocket

julianweise committed:
Introduced notificationService in DashboardView Tests

julianweise committed:
Changed websocket dependency

julianweise committed:
Uninstalled node websocket

Dependency is not necessary anymore. Decided to check whether javascript websocket is supported or not. In case of no support, notification funktionality is not offered and error message is shown to user.

julianweise committed:
Websocket check

Check whether Websocket is supported in environment. If not, NotificationService is not offered and error message is displayed.

julianweise committed:
Introduced binded arguments for callback in ProductView

Binded arguments in function call in order to economize additional argument in notification service.

julianweise committed:
Adapted test for Tab

julianweise committed:
Adapted test for ProductView

Wrote tests for recently indroduced features

julianweise committed:
Adapted tests for DashboardView

Wrote tests for recently indroduced features

julianweise committed:
Adapted test of App.js

julianweise committed:
Fixed typo

Forgot to remove args in previous commit.

julianweise committed:
CodeRefactoring Tab.js

julianweise committed:
Improved code quality of NotificationService

julianweise committed:
Added tests for NotificationService

julianweise commented:
@friedow: Please notice my comment in the  description concerning the coverage decrease.

friedow commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,17 @@
+            <div className={`alert alert-${this.props.data.type}`} role="alert">
+                <table>
+                    <tbody>
+                    <tr>
```
Intendation missing

julianweise replied:
fixed

friedow commented on src/Notification/Notification.scss:
```diff
@@ -0,0 +1,40 @@
+@import "../definitions.scss";
+
+.alert {
```
Let us talk about that.

julianweise replied:
We talked about it. @friedow does not see any need to change anything.

friedow commented on src/NotificationInterface/NotificationService.js:
```diff
@@ -0,0 +1,76 @@
+class NotificationService {
+    
+    constructor(remoteDomain, remotePort, API, notificationCallback) {
+        this.notificationSubscribors = [];
```
In word typo: subscribErs

julianweise replied:
fixed

friedow commented on src/NotificationInterface/NotificationService.js:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    subscribe(entityType, notificationCallback) {
+        const listElement = NotificationService.buildNotificationElement(entityType, notificationCallback);
```
I suggest renaming listElement to subscriber.

julianweise replied:
fixed

friedow commented on src/NotificationInterface/NotificationService.js:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    unsubscribe(entityType, notificationCallback) {
+        const searchElement = this.notificationSubscribors.find((element) => {
```
I suggest renaming element to subscriber.

julianweise replied:
fixed

friedow commented on src/NotificationInterface/NotificationService.js:
```diff
@@ -0,0 +1,76 @@
+            return element.entityOfInterest === entityType && element.notificationCallback === notificationCallback;
+        });
+        const outdatedIndex = this.notificationSubscribors.indexOf(searchElement);
+        if(outdatedIndex > -1) {
```
Instead of indexOf just check searchElement for undefined. 
Array.prototype.find() : Return value: A value in the array if an element passes the test; otherwise, undefined.
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/find

julianweise replied:
Discussed with @friedow, won't fix.

friedow commented on src/NotificationInterface/NotificationService.js:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+    
+    static buildNotificationElement(entityType, notificationCallback) {
```
I suggest renaming buildNotificationElement to buildSubscriber.

julianweise replied:
fixed

julianweise committed:
Fixed typo

julianweise committed:
Improved naming of variable

julianweise committed:
Improved naming of variable

julianweise committed:
Improved naming of method

julianweise committed:
Improved code style

julianweise committed:
Fixed typo in coresponding test

julianweise committed:
Fixed some typos and excluded specific lines from test

Some lines have to be excluded from testing, because our testing framework does not support WebSockets.
